File: backend/services/alpha_vantage.py
"""
Alpha Vantage API Service

MCP-Ready Functions:
- get_top_gainers_losers(): Market scanning (1 API call)
- get_company_overview(): Fundamentals for solvency check
- get_cash_flow(): Operating cash flow for solvency

Rate Limit: 25 requests/day (free tier)
"""

import os
import requests
from typing import Optional
from datetime import datetime


ALPHA_VANTAGE_BASE_URL = "https://www.alphavantage.co/query"


def _get_api_key() -> str:
    """Get Alpha Vantage API key from environment."""
    api_key = os.getenv("ALPHAVANTAGE_API_KEY")
    if not api_key:
        raise ValueError("ALPHAVANTAGE_API_KEY not set in environment")
    return api_key


def get_top_gainers_losers() -> dict:
    """
    Fetch top gainers, losers, and most active from Alpha Vantage.

    MCP Tool: market_scan
    Cost: 1 API call

    Returns:
        dict with keys: top_gainers, top_losers, most_active
        Each contains list of {ticker, price, change_amount, change_percent, volume}
    """
    api_key = _get_api_key()
    params = {
        "function": "TOP_GAINERS_LOSERS",
        "apikey": api_key
    }

    response = requests.get(ALPHA_VANTAGE_BASE_URL, params=params, timeout=30)
    response.raise_for_status()
    data = response.json()

    # Check for API error messages
    if "Information" in data or "Note" in data:
        error_msg = data.get("Information") or data.get("Note")
        raise ValueError(f"Alpha Vantage API error: {error_msg}")

    return {
        "top_gainers": _normalize_movers(data.get("top_gainers", [])),
        "top_losers": _normalize_movers(data.get("top_losers", [])),
        "most_active": _normalize_movers(data.get("most_actively_traded", [])),
        "fetched_at": datetime.utcnow().isoformat()
    }


def get_company_overview(symbol: str) -> dict:
    """
    Fetch company fundamentals for solvency analysis.

    MCP Tool: get_fundamentals
    Cost: 1 API call

    Returns:
        dict with financial metrics
    """
    api_key = _get_api_key()
    params = {
        "function": "OVERVIEW",
        "symbol": symbol.upper(),
        "apikey": api_key
    }

    response = requests.get(ALPHA_VANTAGE_BASE_URL, params=params, timeout=30)
    response.raise_for_status()
    data = response.json()

    if not data or "Symbol" not in data:
        return {"error": f"No data found for {symbol}"}

    return data


def get_cash_flow(symbol: str) -> dict:
    """
    Fetch cash flow statement for detailed solvency check.

    MCP Tool: get_cash_flow
    Cost: 1 API call

    Returns:
        dict with annual and quarterly cash flow reports
    """
    api_key = _get_api_key()
    params = {
        "function": "CASH_FLOW",
        "symbol": symbol.upper(),
        "apikey": api_key
    }

    response = requests.get(ALPHA_VANTAGE_BASE_URL, params=params, timeout=30)
    response.raise_for_status()
    data = response.json()

    if "annualReports" not in data:
        return {"error": f"No cash flow data for {symbol}"}

    return data


def _normalize_movers(movers: list) -> list:
    """Normalize Alpha Vantage mover data to standard format."""
    normalized = []
    for item in movers[:20]:  # Limit to 20
        try:
            change_pct_str = item.get("change_percentage", "0%")
            change_pct = float(change_pct_str.replace("%", ""))

            normalized.append({
                "ticker": item.get("ticker", ""),
                "price": float(item.get("price", 0)),
                "change_amount": float(item.get("change_amount", 0)),
                "change_percent": change_pct,
                "volume": int(item.get("volume", 0))
            })
        except (ValueError, TypeError):
            continue
    return normalized

File: apps/api/src/services/alpha-vantage.ts
/**
 * Alpha Vantage API Service
 *
 * Primary data source for:
 * - Market scanning (TOP_GAINERS_LOSERS)
 * - Company fundamentals (OVERVIEW)
 * - Cash flow statements (CASH_FLOW)
 *
 * Rate Limit: 25 requests/day (free tier)
 */

import type { TickerMovement } from '@stock-researcher/shared';

const ALPHA_VANTAGE_BASE_URL = 'https://www.alphavantage.co/query';

function getApiKey(): string {
  const apiKey = process.env.ALPHAVANTAGE_API_KEY;
  if (!apiKey) {
    throw new Error('ALPHAVANTAGE_API_KEY not set in environment');
  }
  return apiKey;
}

interface AlphaVantageMover {
  ticker: string;
  price: string;
  change_amount: string;
  change_percentage: string;
  volume: string;
}

interface TopGainersLosersResponse {
  top_gainers?: AlphaVantageMover[];
  top_losers?: AlphaVantageMover[];
  most_actively_traded?: AlphaVantageMover[];
  Information?: string;
  Note?: string;
}

interface CashFlowReport {
  operatingCashflow?: string;
  capitalExpenditures?: string;
}

interface CashFlowResponse {
  annualReports?: CashFlowReport[];
  error?: string;
  Information?: string;
  Note?: string;
}

/**
 * Fetch top gainers, losers, and most active from Alpha Vantage.
 * Cost: 1 API call
 */
export async function getTopGainersLosers(): Promise<{
  top_gainers: TickerMovement[];
  top_losers: TickerMovement[];
  most_active: TickerMovement[];
  fetched_at: string;
}> {
  const apiKey = getApiKey();
  const params = new URLSearchParams({
    function: 'TOP_GAINERS_LOSERS',
    apikey: apiKey,
  });

  const response = await fetch(`${ALPHA_VANTAGE_BASE_URL}?${params}`, {
    signal: AbortSignal.timeout(30000),
  });

  if (!response.ok) {
    throw new Error(`Alpha Vantage API error: ${response.status}`);
  }

  const data = await response.json() as TopGainersLosersResponse;

  // Check for API error messages
  if (data.Information || data.Note) {
    throw new Error(`Alpha Vantage API error: ${data.Information || data.Note}`);
  }

  return {
    top_gainers: normalizeMovers(data.top_gainers || []),
    top_losers: normalizeMovers(data.top_losers || []),
    most_active: normalizeMovers(data.most_actively_traded || []),
    fetched_at: new Date().toISOString(),
  };
}

/**
 * Fetch company fundamentals for solvency analysis.
 * Cost: 1 API call
 */
export async function getCompanyOverview(symbol: string): Promise<Record<string, unknown>> {
  const apiKey = getApiKey();
  const params = new URLSearchParams({
    function: 'OVERVIEW',
    symbol: symbol.toUpperCase(),
    apikey: apiKey,
  });

  const response = await fetch(`${ALPHA_VANTAGE_BASE_URL}?${params}`, {
    signal: AbortSignal.timeout(30000),
  });

  if (!response.ok) {
    throw new Error(`Alpha Vantage API error: ${response.status}`);
  }

  const data = await response.json() as Record<string, unknown>;

  if (!data || !data.Symbol) {
    return { error: `No data found for ${symbol}` };
  }

  return data;
}

/**
 * Fetch cash flow statement for detailed solvency check.
 * Cost: 1 API call
 */
export async function getCashFlow(symbol: string): Promise<{
  operatingCashFlow: number | null;
  freeCashFlow: number | null;
  error?: string;
}> {
  const apiKey = getApiKey();
  const params = new URLSearchParams({
    function: 'CASH_FLOW',
    symbol: symbol.toUpperCase(),
    apikey: apiKey,
  });

  const response = await fetch(`${ALPHA_VANTAGE_BASE_URL}?${params}`, {
    signal: AbortSignal.timeout(30000),
  });

  if (!response.ok) {
    throw new Error(`Alpha Vantage API error: ${response.status}`);
  }

  const data = await response.json() as CashFlowResponse;

  if (data.Information || data.Note) {
    return {
      operatingCashFlow: null,
      freeCashFlow: null,
      error: data.Information || data.Note,
    };
  }

  if (!data.annualReports || data.annualReports.length === 0) {
    return {
      operatingCashFlow: null,
      freeCashFlow: null,
      error: `No cash flow data for ${symbol}`,
    };
  }

  const latest = data.annualReports[0];
  const ocfStr = latest.operatingCashflow;
  const capexStr = latest.capitalExpenditures;

  const ocf = ocfStr && ocfStr !== 'None' ? parseFloat(ocfStr) : null;
  const capex = capexStr && capexStr !== 'None' ? parseFloat(capexStr) : 0;
  const fcf = ocf !== null ? ocf - Math.abs(capex) : null;

  return {
    operatingCashFlow: ocf,
    freeCashFlow: fcf,
  };
}

/**
 * Normalize Alpha Vantage mover data to standard format.
 */
function normalizeMovers(movers: AlphaVantageMover[]): TickerMovement[] {
  const normalized: TickerMovement[] = [];

  for (const item of movers.slice(0, 20)) {
    try {
      const changePctStr = item.change_percentage || '0%';
      const changePct = parseFloat(changePctStr.replace('%', ''));

      normalized.push({
        ticker: item.ticker || '',
        price: parseFloat(item.price) || 0,
        change_amount: parseFloat(item.change_amount) || 0,
        change_percent: changePct,
        volume: parseInt(item.volume) || 0,
      });
    } catch {
      // Skip invalid entries
      continue;
    }
  }

  return normalized;
}

// Export as module
export const alphaVantage = {
  getTopGainersLosers,
  getCompanyOverview,
  getCashFlow,
};
